File: acme-admin/src/main/java/com/acme/admin/service/security/PermissionCatalog.java
package com.acme.admin.service.security;

import java.util.*;

public final class PermissionCatalog {
  private PermissionCatalog() {}

  public record PermissionSpec(String code, String description) {}

  public static List<PermissionSpec> specs() {
    return List.of(
      new PermissionSpec("users.read", "List/read users"),
      new PermissionSpec("users.write", "Create/update users"),
      new PermissionSpec("roles.read", "List/read roles"),
      new PermissionSpec("roles.write", "Create/update roles"),
      new PermissionSpec("finance.read", "Read finance data"),
      new PermissionSpec("finance.write", "Write finance data")
    );
  }

  public static Set<String> allCodes() {
    final Set<String> out = new LinkedHashSet<>();
    for (final PermissionSpec s : specs()) out.add(s.code());
    return Collections.unmodifiableSet(out);
  }
}

File: acme-admin-dotnet/Security/PermissionCatalog.cs
namespace Acme.Admin.Api.Security;

public static class PermissionCatalog
{
    public sealed record PermissionSpec(string Code, string Description);

    public static IReadOnlyList<PermissionSpec> Specs() =>
    [
        // Canonical .NET policy permission catalog.
        new("users.read", "List/read users"),
        new("users.write", "Create/update users"),
        new("roles.read", "List/read roles"),
        new("roles.write", "Create/update roles"),
        new("taxes.read", "List/read taxes"),
        new("taxes.write", "Create/update/delete taxes"),
        new("orders.read", "List/read orders"),
        new("orders.write", "Create/update orders"),
        new("procurement.read", "List/read procurement data"),
        new("procurement.write", "Create/update procurement data"),
        new("finance.read", "Read finance data"),
        new("finance.write", "Write finance data"),
        new("hr.read", "Read HR data"),
        new("hr.write", "Write HR data"),
        new("catalog.read", "Read catalog"),
        new("catalog.write", "Write catalog"),
        new("billing.read", "Read billing data"),
        new("billing.create", "Create billing sessions"),
        new("billing.pay", "Pay billing intents")
    ];

    public static ISet<string> AllCodes()
    {
        return Specs().Select(x => x.Code).ToHashSet(StringComparer.Ordinal);
    }
}
